### DISCUSSION TITLE
How can i get the list of geometry of a shape

### DISCUSSION THREAD

**Discussion by Daoudou** [openscenegraph/OpenSceneGraph#1212] [Q&A] [answered]
Hello

I want to know how can i get the osg::Geometry of a shape ?? like osg::Cone for example.

Thanks 
SOUFFOU

**Comment by robertosfield** [comment, marked as answer]
The osg::Shape implementations don't have a osg::Geometry so you can "get" one from it.  The osg::Shape are just primitive shape definitions that don't have any additional functionality built-in.

In order to be render a shape you need to use a osg::ShapeDrawable, but this isn't what you asked, perhaps what you meant.  Look at the osgshape example.

**Comment by Daoudou** [comment]
Hello 
thanks for you're answer.

I want to use the project osgModeling with osg::ShapeDrawable but i need osg::Geometry for that, i start with TriangleFunctor for duplicate the facets of a cone to do a Boolean Operator with a Box but i have this result : 

![Capture2](https://user-images.githubusercontent.com/98310378/227944006-c359cf90-3250-4e2c-a073-5bab41224acd.PNG)

my code : 

    osg::ref_ptr<osg::Geometry> createConeDirtyWay() {

    osg::ref_ptr<osg::Geometry> geom = new osg::Geometry;
    osg::ref_ptr<osg::Vec3Array> vertices = new osg::Vec3Array(arrayVecCone.size());
    
    for (unsigned int i = 0; i < arrayVecCone.size(); i++)
    {
        (*vertices)[i].set(arrayVecCone.at(i));
    }

    geom->setVertexArray(vertices.get());
    geom->addPrimitiveSet(new osg::DrawArrays(GL_TRIANGLES, 0, arrayVecCone.size()));

    return geom;
    }

    osg::ref_ptr<osg::Geometry> createCubeDirtyWay() {

    osg::ref_ptr<osg::Geometry> geom = new osg::Geometry;
    osg::ref_ptr<osg::Vec3Array> vertices = new osg::Vec3Array(arrayVecCube.size());


    for (unsigned int i = 0; i < arrayVecCube.size(); i++)
    {
        (*vertices)[i].set(arrayVecCube.at(i));
    }

    geom->setVertexArray(vertices.get());
    geom->addPrimitiveSet(new osg::DrawArrays(GL_TRIANGLES, 0, arrayVecCube.size()));

    return geom;
    }

// i use vector array to get every Vec3 of facets of Box and Cone

`struct FaceCollectorCone
    {
     void operator()(const osg::Vec3& v1, const osg::Vec3& v2,
        const osg::Vec3& v3, bool)
    {
        arrayVecCone.push_back(v1);
        arrayVecCone.push_back(v2);
        arrayVecCone.push_back(v3);
    }
   };

 struct FaceCollectorCube
 {
    void operator()(const osg::Vec3& v1, const osg::Vec3& v2,
        const osg::Vec3& v3, bool)
    {
        arrayVecCube.push_back(v1);
        arrayVecCube.push_back(v2);
        arrayVecCube.push_back(v3);
    }
 };

// in the main fonction : 
  osg::ref_ptr<osg::Cone> coneModel = new osg::Cone(osg::Vec3f(0.,0.,0.),5,15);
  osg::ref_ptr<osg::ShapeDrawable> shapeCone = new osg::ShapeDrawable;

  shapeCone->setShape(coneModel);

  osg::TriangleFunctor<FaceCollectorCone> functor;
  shapeCone->accept(functor);

  osg::ref_ptr<osg::Box> cubeModel = new osg::Box(osg::Vec3f(0., 0., 9.), 5);
  osg::ref_ptr<osg::ShapeDrawable> shapeCube = new osg::ShapeDrawable;

  shapeCube->setShape(cubeModel);

  osg::TriangleFunctor<FaceCollectorCube> functorCube;
  shapeCube->accept(functorCube);`
    

it's look like the Cone have too much Vec3 in my vector array.

**Comment by robertosfield** [comment]
I am not familiar with the osgModelling library so can't help with this.

Looking at osg::ShapeDrawable in the OSG-3.6 branch, it subclasses from osg::Geometry so you should be able to just query it directly for it's vertex and primitive data.
